Dockerfile

```diff
@@ -0,0 +1,22 @@
+FROM python:3.7.7-slim
+
+# Create virtualenv and add to path.
+ENV VIRTUAL_ENV=/opt/venv
+RUN python3 -m venv $VIRTUAL_ENV
+ENV PATH="$VIRTUAL_ENV/bin:$PATH"
+
+# System dependencies
+RUN apt-get update && apt-get install -y build-essential && rm -rf /var/lib/apt/lists/*
+
+WORKDIR /opt/app
+
+# Install Python requirements. README.md is required as the setup.py file
+# refers to it.
+COPY README.md .
+COPY setup.py .
+RUN pip install -e .[dev,test]
+
+# Run py.test against current dir by default but allow custom args to be passed
+# in.
+ENTRYPOINT ["py.test"]
+CMD [""]
```

Comment:
This makes it easy to run tests but a bit of a pain to run, say, `isort`. You have to do this:
```bash
$ docker run --rm -v `pwd`:/opt/app --entrypoint=isort xocto -rc
```

Better to have a more general-purpose run-something-within-the-venv image rather than a specialist run-tests-within-the-venv?

Reply:
I may get into trouble for saying this, but I think the Docker way of doing things is quite clear - they expect you to install everything using a Dockerfile and have a single command that the container runs and that is it. It isn't designed to run different commands, so if you want to do that then it will feel painful and I think that is a design decision on their part. 

Using docker to run all the other parts of your dev environment (databases etc) is a great idea, but changing the code inside a container just to get the Python environment you want? Just use a virtualenv!

All of that said, _if_ you want to use Docker like this then you can use `docker exec` because you don't need a new container (exec = existing container, run = new container).

Reply:
That depends on how you see this being used... I would usually say that an image meant to run a program should have an entrypoint, one meant to contain a dev environment should not.

If you're looking to make a single image run multiple programs, you're doing it kinda wrong :-P I would suggest not specifying the entrypoint or command in the base image, then extend that image in a series of named later stages:
```
FROM python:3.7.7-slim as base

...

FROM base as pytest

# Run py.test against current dir by default but allow custom args to be passed
# in.
ENTRYPOINT ["py.test"]
CMD [""]

FROM base as black

# Run black against current dir by default but allow custom args to be passed
# in.
ENTRYPOINT ["black"]
CMD [""]
```

You can then build these using `docker build -t xocto/pytest --target pytest . && docker build -t xocto/black --target black .`, and basically the whole image should be shared.

Reply:
@dooferlad You are definitely right about the docker images in general.  However, I do kinda see the point of having a perfectly sandboxed and consistent testing/linting environment that you just inject your code into.  This avoids things like global linting configuration accidentally impacting your linting for a tightly controlled individual project.  So I can see the point.

Reply:
Thanks both - that's interesting. It doesn't really matter for this project - I was just using it as a sandbox.

I like @scalen's idea of creating multiple single-process images from the same base. I find it a bit weird that building the `black` image still runs all the commands from the `pytest` image but I guess it does not harm when the leaf images just contain `ENTRYPOINT` and `CMD` commands (which don't create new layers).

Reply:
> but changing the code inside a container just to get the Python environment you want? Just use a virtualenv!

Having a dev image like this might make sense when building the virtualenv is heavy-weight (ie lots of compilation). I thought that was the case here due to pandas but later realised I could upgrade and install a wheel instead.

Reply:
> I find it a bit weird that building the `black` image still runs all the commands from the `pytest` image but I guess it does not harm when the leaf images just contain `ENTRYPOINT` and `CMD` commands (which don't create new layers).

It's true that it doesn't matter much as long as the pytest image is on the same box, but you're right, it could be smaller and still based on the same stack of layers.  You'd just need to break up the commands in the `base` image into smaller stages, and base the `black` image off a stage before the layers relating to `pytest` 🤷